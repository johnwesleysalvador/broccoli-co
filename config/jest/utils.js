import React from 'react';
import { shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import theme from '../../src/themes/default';

const shallowWithStore = (component, store) => {
    const context = {
        store,
    };
    const themeContext = shallow(<ThemeProvider theme={theme} />)
        .instance()
        .getChildContext();

    return shallow(component, {
        context: { ...context, ...themeContext },
        childContextTypes: ThemeProvider.childContextTypes,
    });
};

export default { shallowWithStore };
