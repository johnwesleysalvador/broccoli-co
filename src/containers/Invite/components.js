import Paper from '@material-ui/core/Paper';
import Icon from '@material-ui/core/Icon';
import styled from 'styled-components';

import Button from 'components/Button';

import bg from 'assets/bg.jpg';

export const Wrapper = styled.div`
    display: flex;

    height: 100%;
`;

export const Container = styled(Paper)`
    max-width: 700px;
    margin: auto;
    display: flex;
    flex-wrap: wrap;
`;

export const Welcome = styled.div`
    padding: 80px;
    flex: 1;
    background: linear-gradient(#3f51b5, #3f51b5d4), url(${bg}) no-repeat center;
    background-size: cover;
    position: relative;
`;

export const SubHeading = styled.div`
    font-style: italic;
`;

export const WelcomeBody = styled.div`
    z-index: 1;
    color: white;
    text-align: center;
`;

export const InviteForm = styled.form`
    padding: 40px;
    flex: 1.5;
    display: flex;
    flex-direction: column;
`;

export const SendIcon = styled(Icon)`
    margin-left: 10px;
`;

export const SendButton = styled(Button)`
    && {
        margin-top: 20px;
    }
`;
