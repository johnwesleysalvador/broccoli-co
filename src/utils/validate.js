export const required = value => !value && 'Required';

export const email = value =>
    value && !value.match(/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/i) && 'Invalid email';

export const minLength = min => value =>
    value && value.length < min && `Must have at least ${min} characters or more`;

export const minLength3 = minLength(3);

// export const minLength = (value) =>
