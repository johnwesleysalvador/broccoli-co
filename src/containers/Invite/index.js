import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { reduxForm, Field } from 'redux-form';
import { connect } from 'react-redux';
import Typography from '@material-ui/core/Typography';

import Dialog from 'components/Dialog';
import Input from 'components/Input';

import { sendInvite, clearState } from 'ducks/invite';
import { required, minLength3, email as emailValidation } from 'utils/validate';

import {
    Wrapper,
    Container,
    Welcome,
    InviteForm,
    SendIcon,
    SendButton,
    WelcomeBody,
    SubHeading,
} from './components';

class Invite extends PureComponent {
    handleSubmit = ({ name, email }) => {
        const { dispatch } = this.props;

        return dispatch(sendInvite({ name, email }));
    };

    handleDialogClose = () => {
        const {
            dispatch,
            reset,
            invite: { status },
        } = this.props;

        if (status === 200) reset();
        dispatch(clearState());
    };

    render() {
        const {
            invite: { status, error, isLoading },
            handleSubmit,
        } = this.props;
        return (
            <Wrapper>
                <Dialog open={status === 200} title="All done!" onClose={this.handleDialogClose}>
                    You will be one of the first to experience Brocoli &amp; Co. when we launch
                </Dialog>
                <Dialog open={status === 400} title="Sorry!" onClose={this.handleDialogClose}>
                    {error}
                </Dialog>
                <Container>
                    <Welcome>
                        <WelcomeBody>
                            <Typography variant="display1" gutterBottom color="inherit">
                                A better way to enjoy every day.
                            </Typography>
                            <SubHeading variant="subheading" color="inherit">
                                Be the first to know when we launch.
                            </SubHeading>
                        </WelcomeBody>
                    </Welcome>
                    <InviteForm onSubmit={handleSubmit(this.handleSubmit)}>
                        <Typography variant="headline" gutterBottom>
                            Request an invite
                        </Typography>
                        <Field
                            component={Input}
                            name="name"
                            label="Full name"
                            validate={[required, minLength3]}
                        />
                        <Field
                            component={Input}
                            name="email"
                            label="Email"
                            validate={[required, emailValidation]}
                            type="email"
                        />
                        <Field
                            component={Input}
                            name="confirmEmail"
                            label="Confirm email"
                            validate={[required, emailValidation]}
                            type="email"
                        />

                        <SendButton
                            type="submit"
                            variant="raised"
                            color="primary"
                            {...isLoading && { disabled: true }}
                        >
                            {isLoading ? 'Sending...' : 'Send'}
                            <SendIcon>send</SendIcon>
                        </SendButton>
                    </InviteForm>
                </Container>
            </Wrapper>
        );
    }
}

Invite.propTypes = {
    dispatch: PropTypes.func.isRequired,
    reset: PropTypes.func.isRequired,
    invite: PropTypes.shape({
        status: PropTypes.number,
        error: PropTypes.string,
    }).isRequired,
    handleSubmit: PropTypes.func.isRequired,
};

const validate = ({ email, confirmEmail }) => {
    const error = {};

    if (email !== confirmEmail) {
        error.confirmEmail = 'Emails do not match!';
    }

    return error;
};

const mapStateToProps = ({ invite }) => ({
    invite,
});

export default connect(mapStateToProps)(reduxForm({ form: 'inviteForm', validate })(Invite));
