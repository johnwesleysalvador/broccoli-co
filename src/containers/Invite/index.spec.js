import React from 'react';
import { createMockStore } from 'redux-test-utils';
import 'jest-styled-components';
import utils from '../../../config/jest/utils';

import Invite from '.';

describe('/containers/Invite', () => {
    test('should render successfully', () => {
        const testState = {
            invite: { isLoading: false },
        };
        const store = createMockStore(testState);
        const component = utils.shallowWithStore(<Invite />, store);

        expect(component.dive()).toMatchSnapshot();
    });
    test('should simulate send invite action', () => {
        const testState = {
            invite: { isLoading: false, status: 0 },
        };
        const payload = { email: 'john@doe.com', name: 'John Doe' };
        const store = createMockStore(testState);
        const component = utils.shallowWithStore(<Invite />, store);
        // Dive until Invite component is found
        const instance = component
            .dive()
            .dive()
            .dive()
            .dive()
            .instance();

        instance.handleSubmit(payload);
        expect(store.isActionDispatched({
            type: 'invite/SEND_INVITE',
            payload,
        })).toEqual(true);
    });
    test('should simulate closing a dialog', () => {
        const testState = {
            invite: { isLoading: false, status: 0 },
        };
        const store = createMockStore(testState);
        const component = utils.shallowWithStore(<Invite />, store);
        // Dive until Invite component is found
        const instance = component
            .dive()
            .dive()
            .dive()
            .dive()
            .instance();

        instance.handleDialogClose();
        expect(store.isActionDispatched({
            type: 'invite/CLEAR_STATE',
        })).toEqual(true);
    });
});
