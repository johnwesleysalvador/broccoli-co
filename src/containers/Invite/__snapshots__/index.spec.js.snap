// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/containers/Invite should render successfully 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    dispatch={[Function]}
    invite={
        Object {
            "isLoading": false,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "destroyOnUnmount": true,
      "dispatch": [Function],
      "enableReinitialize": false,
      "forceUnregisterOnUnmount": false,
      "form": "inviteForm",
      "getFormState": [Function],
      "initialValues": undefined,
      "invite": Object {
        "isLoading": false,
      },
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldError": [Function],
      "shouldValidate": [Function],
      "shouldWarn": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
      "updateUnregisteredFields": false,
      "validate": [Function],
    },
    "ref": [Function],
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "destroyOnUnmount": true,
        "dispatch": [Function],
        "enableReinitialize": false,
        "forceUnregisterOnUnmount": false,
        "form": "inviteForm",
        "getFormState": [Function],
        "initialValues": undefined,
        "invite": Object {
          "isLoading": false,
        },
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldError": [Function],
        "shouldValidate": [Function],
        "shouldWarn": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
        "updateUnregisteredFields": false,
        "validate": [Function],
      },
      "ref": [Function],
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "childContextTypes": Object {
      "__styled-components__": [Function],
      "__styled-components__next__": [Function],
    },
    "context": Object {
      "__styled-components__": [Function],
      "__styled-components__next__": Object {
        "getTheme": [Function],
        "subscribe": [Function],
        "unsubscribe": [Function],
      },
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
